File: api/app/schemas/rag.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    file_name: str
    content_type: str
    chunk_count: int
    entity_count: int
    relationship_count: int
    uploaded_by_user_id: int | None
    uploaded_by_name: str | None
    created_at: datetime
    updated_at: datetime


class DocumentChunkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_id: int
    document_title: str
    chunk_index: int
    content: str
    token_count: int
    created_at: datetime
    updated_at: datetime


class ChatbotQuestion(BaseModel):
    question: str = Field(min_length=1, max_length=1000)


class ChatbotSource(BaseModel):
    document_id: int
    document_title: str
    chunk_id: int
    chunk_index: int
    score: float
    preview: str


class ChatbotMatchedEntity(BaseModel):
    id: int
    name: str
    entity_type: str
    description: str


class ChatbotAnswer(BaseModel):
    answer: str
    sources: list[ChatbotSource]
    confidence_score: float = 0.0
    matched_entities: list[ChatbotMatchedEntity] = Field(default_factory=list)
    fallback_reason: str | None = None

File: web/types/rag.ts
export type DocumentRecord = {
  id: number;
  title: string;
  file_name: string;
  content_type: string;
  chunk_count: number;
  entity_count: number;
  relationship_count: number;
  skipped_duplicate_chunks: number;
  page_count: number;
  extracted_character_count: number;
  average_chunk_length: number;
  max_chunk_length: number;
  conflict_warning: string | null;
  processing_status: string;
  processing_error: string | null;
  uploaded_by_user_id: number | null;
  uploaded_by_name: string | null;
  created_at: string;
  updated_at: string;
};

export type DocumentChunk = {
  id: number;
  document_id: number;
  document_title: string;
  chunk_index: number;
  content: string;
  token_count: number;
  created_at: string;
  updated_at: string;
};

export type ChatbotSource = {
  document_id: number;
  document_title: string;
  chunk_id: number;
  chunk_index: number;
  score: number;
  preview: string;
};

export type ChatbotAnswer = {
  answer: string;
  sources: ChatbotSource[];
  confidence_score: number;
  matched_entities?: {
    id: number;
    name: string;
    entity_type: string;
    description: string;
  }[];
  fallback_reason?: string | null;
};
